File: HW3/hw3_problem2.py
"""
MAE 6760 Model Based Estimation
Cornell University
M Campbell

Homework #3
Problem #2: Kalman Filter
    Two aircraft tracking estimator in 2D as they turn
    uses the following Python functions:
        plot_openloop.py
        plot_estimator.py
"""

import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import cont2discrete
from scipy.stats import chi2
from scipy.linalg import sqrtm, block_diag

from plot_estimator import plot_estimator

np.random.seed(101)

# ============================================================
# 2D Aircraft Open Loop Plot
# ============================================================

def plot_openloop_2D_aircraft(x_A, x_B=None):

    plt.figure(figsize=(8,6))

    plt.plot(x_A[0,:], x_A[2,:], 'b-')
    plt.plot(x_A[0,0], x_A[2,0], 'b>')
    plt.plot(x_A[0,-1], x_A[2,-1], marker='*', color='b')

    if x_B is not None:
        plt.plot(x_B[0,:], x_B[2,:], 'r:')
        plt.plot(x_B[0,0], x_B[2,0], 'r>')
        plt.plot(x_B[0,-1], x_B[2,-1], marker='*', color='r')
        plt.legend(["Aircraft A","Aircraft B"])
    else:
        plt.legend(["Aircraft A"])

    plt.xlabel("East")
    plt.ylabel("North")
    plt.grid(True)


# ============================================================
# Aircraft A Model
# ============================================================

# continuous time model for aircraft A
Omega_A = 0.045
nx = 4
A_A = np.array([[0,1,0,0],
                [0,0,0,-Omega_A],
                [0,0,0,1],
                [0,Omega_A,0,0]])

B_A = np.array([[0,0],[1,0],[0,0],[0,1]])
H_A = np.array([[1,0,0,0],
                [0,0,1,0]])

# discrete time model for aircraft A
dt = 0.5
F_A, G_A, _, _, _ = cont2discrete((A_A,B_A,np.eye(4),0), dt)

Q_A = 10*np.array([[2.0,0.05],[0.05,0.5]])
R_A = np.array([[20,0.05],[0.05,20]])

tvec = np.arange(0,100+dt,dt)
nk = len(tvec)

x0_A = np.array([0,85*np.cos(np.pi/4),0,-85*np.sin(np.pi/4)])

# simulate aircraft A
w_A = sqrtm(Q_A) @ np.random.randn(2,nk)
x_A = np.zeros((4,nk))
x_A[:,0] = x0_A

for k in range(nk-1):
    x_A[:,k+1] = F_A @ x_A[:,k] + G_A @ w_A[:,k]

z_A = H_A @ x_A + sqrtm(R_A) @ np.random.randn(2,nk)

plot_openloop_2D_aircraft(x_A)
plt.title("2D trajectory for aircraft A")

# ============================================================
# (a) Baseline Kalman Filter
# ============================================================
# goal: KF to track/localize aircraft A
# 
# use model for aircraft A for this problem
# use measurements z_A for this problem
# 

## YOUR CODE HERE

fig, axs = plt.subplots(1,2,figsize=(16,6))
plot_estimator(tvec,xhatu[0,:],Pu[0,0,:],x_A[0,:],plot_type='error',z=z_A[0,:],ax=axs[0])
axs[0].set_ylabel("North error estimate $e_N(t)$")
plot_estimator(tvec,xhatu[2,:],Pu[2,2,:],x_A[2,:],plot_type='error',z=z_A[1,:],ax=axs[1])
axs[1].set_ylabel("East error estimate $e_E(t)$")
plt.suptitle("(a) Baseline Kalman Filter for aircraft A", fontweight='bold')
plt.tight_layout()


# ============================================================
# (b) 95% Measurement Gating
# ============================================================
# goal: KF to track/localize aircraft A with measurement gate
# 
# use model for aircraft A for this problem
# use measurements z_A for this problem
# 
# variables to generate for display/plots:
#  Nrej: number of measurement rejections (subset of total nk)
#  Trej: times tvec[k+1] when msmt is rejected 
#  Erej: error z[:,k+1]-H*x_A(:,k+1] when msmt is rejected

# msmt gating parameters
Lam0 = chi2.ppf(0.95,2)
Lam = np.zeros(nk)

Nrej = 0
Irej = []

## YOUR CODE HERE

# uncomment lines below to output the percent of msmts rejected
# print("(b) percent msmts rejected:",Nrej/nk*100)

fig, axs = plt.subplots(1,2,figsize=(16,6))

plot_estimator(tvec,xhatu[0,:],Pu[0,0,:],x_A[0,:],plot_type='error',z=z_A[0,:],ax=axs[0])
# uncomment the line below to add msmt rejections to the plot
# axs[0].plot(Trej, Erej[0,:], 'mo', fillstyle='none', label='Msmt Rejection')
axs[0].set_ylabel("North error estimate $e_N(t)$")
axs[0].legend()

plot_estimator(tvec,xhatu[2,:],Pu[2,2,:],x_A[2,:],plot_type='error',z=z_A[1,:],ax=axs[1])
# uncomment the line below to add msmt rejections to the plot
# axs[1].plot(Trej, Erej[1,:], 'mo', fillstyle='none', label='Msmt Rejection')
axs[1].set_ylabel("East error estimate $e_E(t)$")
axs[1].legend()
plt.suptitle("(b) Kalman Filter with 95% msmt gating for aircraft A", fontweight='bold')
plt.tight_layout()


# ============================================================
# (c) Process Noise Tuning + Filter Integrity
# ============================================================
# goal: use hypothesis tests to tune process noise covariance Q 
# 
# use model for aircraft A for this problem
# load simulated measurements from tile (these are for aircraft A, but labeled z_c for Part c
# use measurements z_c for this problem
# use states x_c for the plots on problem
# 
# variables to generate for display/plots:
#  Lam: innovations test statistic lambda (1 x nk)
#  LamF: Kalman Filter test statistic lambda (1 x nk)
#  Nrej: number of measurement rejections (subset of total nk)
#  Trej: times tvec[k+1] when msmt is rejected 
#  Erej: error z[:,k+1]-H*x_c[:,k+1] when msmt is rejected
#  NFrej: number of times the filter is inconsistent (subset of total nk)
#  TFrej: times tvec[k+1] when the filter is inconsistent 
#  EFrej: error z[:,k+1]-H*x_c[:,k+1] when the filter is inconsistent

#load measurement vector z_c and state x_c
data = np.load("ACdata.npz")
z_c = data["z_c"]
x_c = data["x_c"]

Nrej = 0
Irej = []
NFrej = 0
IFrej = []

# msmt gating parameters
Lam0 = chi2.ppf(0.95,2)
Lam = np.zeros(nk)

# filter consistency parameters
Lam = np.zeros(nk)
LamF = np.zeros(nk)
win=10
Blow = chi2.ppf(0.025,10*2)/win
Bhigh = chi2.ppf(0.975,10*2)/win

## YOUR CODE HERE

# uncomment lines below to output the percent of msmts rejected
# print("(c) percent msmts rejected:",Nrej/nk*100)
# print("(c) percent time KF inconsistent:",NFrej/nk*100)

fig, axs = plt.subplots(1,2,figsize=(16,6))

plot_estimator(tvec,xhatu[0,:],Pu[0,0,:],x_c[0,:],plot_type='error',z=z_c[0,:],ax=axs[0])
# uncomment the line below to add msmt rejections to the plot
# axs[0].plot(Trej, Erej[0,:], 'mo', fillstyle='none', label='Msmt Rejection')
axs[0].set_ylabel("North error estimate $e_N(t)$")
axs[0].legend()
plot_estimator(tvec,xhatu[2,:],Pu[2,2,:],x_c[2,:],plot_type='error',z=z_c[1,:],ax=axs[1])
# uncomment the line below to add msmt rejections to the plot
# axs[1].plot(Trej, Erej[1,:], 'mo', fillstyle='none', label='Msmt Rejection')
axs[1].set_ylabel("East error estimate $e_E(t)$")
axs[1].legend()
plt.suptitle("(c): Kalman Filter with Msmt Gate and Filter Integrity for aircraft A", fontweight='bold')
plt.tight_layout()

plt.figure(figsize=(8,6))
plt.plot(tvec,Lam,'m:',label=r'inn test statistic $\lambda$')
plt.plot(tvec[win+1:],LamF[win+1:],'-',color='purple',label=r'KF test statistic $\lambda_{10}^{KF}$')
plt.plot(tvec,Blow*np.ones(nk),'b--',label='lower bound')
plt.plot(tvec,Bhigh*np.ones(nk),'b--',label='upper bound')
plt.ylabel('filter integrity')
plt.xlabel('time $t$ (sec)')
plt.legend()
plt.title("(c) Kalman Filter integrity test statistic", fontweight='bold')


# ============================================================
# (d) Joint Kalman Filter (Aircraft A & B)
# ============================================================
# goal: estimate X=[x_A;x_B] given Z=[z_A;z_B]
# 
# use both models for aircraft A and B for this problem
# use measurements Z_A and Z_B for this problem

# continuous time model for aircraft B
Omega_B = -0.045
A_B = np.array([[0,1,0,0],
                [0,0,0,-Omega_B],
                [0,0,0,1],
                [0,Omega_B,0,0]])
B_B = np.array([[0,0],[1,0],[0,0],[0,1]])

# discrete time model for aircraft B
F_B, G_B, _, _, _ = cont2discrete((A_B,B_B,np.eye(nx),0), dt)
Q_B = Q_A
R_B = R_A

# simulate aircraft B
x0_B = np.array([4000,85*np.cos(np.pi/4),3200,-85*np.sin(np.pi/4)])

x_B = np.zeros((4,nk))
x_B[:,0]=x0_B
for k in range(nk-1):
    x_B[:,k+1] = F_B@x_B[:,k] + G_B@w_A[:,k]

z_B = H_A@x_B + sqrtm(R_B)@np.random.randn(2,nk)

plot_openloop_2D_aircraft(x_A,x_B)

## YOUR CODE HERE

fig, axs = plt.subplots(1,2,figsize=(16,6))
plot_estimator(tvec,xhatu[0,:],Pu[0,0,:],x_A[0,:],plot_type='error',z=z_A[0,:],ax=axs[0])
axs[0].set_ylabel("North error estimate $e_N(t)$")
axs[0].legend()
plot_estimator(tvec,xhatu[2,:],Pu[2,2,:],x_A[2,:],plot_type='error',z=z_A[1,:],ax=axs[1])
axs[1].set_ylabel("East error estimate $e_E(t)$")
axs[1].legend()
plt.suptitle("(d): Joint Kalman Filter: aircraft A errors", fontweight='bold')
plt.tight_layout()

fig, axs = plt.subplots(1,2,figsize=(16,6))
plot_estimator(tvec,xhatu[nx+0,:],Pu[nx+0,nx+0,:],x_B[0,:],plot_type='error',z=z_B[0,:],ax=axs[0])
axs[0].set_ylabel("North error estimate $e_N(t)$")
axs[0].legend()
plot_estimator(tvec,xhatu[nx+2,:],Pu[nx+2,nx+2,:],x_B[2,:],plot_type='error',z=z_B[1,:],ax=axs[1])
axs[1].set_ylabel("East error estimate $e_E(t)$")
axs[1].legend()
plt.suptitle("(d): Joint Kalman Filter: aircraft B errors", fontweight='bold')
plt.tight_layout()


# ============================================================
# (e) Joint Kalman Filter with Relative Range Sensor
# ============================================================
# goal: estimate X=[X_A;X_B] given Z=[z_A;z_B]
#       --> repeat (d), but for a 2D range measurement
# 
# use both models for aircraft A and B for this problem
# use measurements Zr for this problem

Rr = np.array([[10,0.15],[0.15,10]])
Hr = np.array([[1,0,0,0,-1,0,0,0],
               [0,0,1,0,0,0,-1,0]])

Zr = np.vstack((x_A[0,:]-x_B[0,:],
                x_A[2,:]-x_B[2,:])) + sqrtm(Rr)@np.random.randn(2,nk)


## YOUR CODE HERE


fig, axs = plt.subplots(1,2,figsize=(16,6))
plot_estimator(tvec,xhatu[0,:],Pu[0,0,:],x_A[0,:],plot_type='error',z=z_A[0,:],ax=axs[0])
axs[0].set_ylabel("North error estimate $e_N(t)$")
axs[0].legend()
plot_estimator(tvec,xhatu[2,:],Pu[2,2,:],x_A[2,:],plot_type='error',z=z_A[1,:],ax=axs[1])
axs[1].set_ylabel("East error estimate $e_E(t)$")
axs[1].legend()
plt.suptitle("(e): Joint KF with Relative Range: aircraft A errors", fontweight='bold')
plt.tight_layout()

fig, axs = plt.subplots(1,2,figsize=(16,6))
plot_estimator(tvec,xhatu[nx+0,:],Pu[nx+0,nx+0,:],x_B[0,:],plot_type='error',z=z_B[0,:],ax=axs[0])
axs[0].set_ylabel("North error estimate $e_N(t)$")
axs[0].legend()
plot_estimator(tvec,xhatu[nx+2,:],Pu[nx+2,nx+2,:],x_B[2,:],plot_type='error',z=z_B[1,:],ax=axs[1])
axs[1].set_ylabel("East error estimate $e_E(t)$")
axs[1].legend()
plt.suptitle("(e): Joint KF with Relative Range: aircraft B errors", fontweight='bold')
plt.tight_layout()

plt.show()

File: HW3/hw3_problem2.m
%% MAE 6760 Model Based Estimation
% Cornell University
% M Campbell
%
% Homework #3
% Problem #2: Kalman Filter
%   Two aircraft tracking estimator in 2D as they turn
%   Plotting uses plot_openloop.m and plot_estimator.m
%
clear all; close all
MAE6760startup; %can adjust font size, figure size at the bottom of this script
global MCcolors; %define colors as global with access 
rng(101);
 
%% Define system model (CT and DT) for aircraft A 
%continuous time model for aircraft A
Omega_A = 0.045; %turn velocity for A in rad/s
%state vector: [East pos, East vel, North pos, North vel]
nx=4;
A_A=[0 1 0 0;0 0 0 -Omega_A;0 0 0 1;0 Omega_A 0 0]; 
ix=1;iy=3; %indices of the 2D position
%input forces (East, North)
B_A=[0 0;1 0;0 0;0 1]; 
nw=2;
%discrete time model 
dt = 0.5; %sec
[F_A,G_A]=c2d(A_A,B_A,dt);
% sensor output models: 2D position states
H_A = [1 0 0 0; 
       0 0 1 0];
nz=2;
% build discrete time model for aircraft A
DTsys_A = ss(F_A,G_A,H_A,zeros(nz,nw),dt); 
%model directional wind disturbances
Q_A = 10*[2.0 0.05;    
         0.05 0.5];

%% simulate aircraft A with process noise and plot 
tvec = 0:dt:100;nk=length(tvec);
x0_A = [0,85*cos(pi/4),0,-85*sin(pi/4)]';
w_A = sqrtm(Q_A)*randn(nw,nk);
[~,~,x_A] = lsim(DTsys_A,w_A,tvec,x0_A);x_A=x_A'; %make into row vector
z_A_clean = H_A*x_A;
%
%simulate measurements for aircraft A from a tracking station
R_A = [20 0.05; 
      0.05 20];
v_A = sqrtm(R_A)*randn(nz,nk);
z_A = H_A*x_A+v_A;

%plot trajectory for aircraft A for some context
figs(1)=figure('Position',[100 100 800 600]);
plot_openloop_2Daircraft(x_A);
title('2D trajectory for aircraft A')


%% Part a) for aircraft A: generate measurements and create a Kalman Filter
%goal: KF to track/localize aircraft A
%
%use model DTsys_A for this problem
%use measurements z_A for this problem
%

F=DTsys_A.A;G=DTsys_A.B;H=H_A;
nx=length(F);
Q=Q_A;R=R_A;z=z_A;

%%YOUR CODE HERE

%Generate plot of East and North position error estimates and 2-sigma bounds
%assumes t is (1 x nk), xhatu is (nx x nk), Pu is (nx x nx x nk), xtrue is (nx x nk) 
%
figs(2)=figure;
ti(2)=tiledlayout(1,2,'TileSpacing','compact','Padding','tight');
nexttile;
plot_estimator(tvec,xhatu(ix,:),Pu(ix,ix,:),x_A(ix,:),'error',z_A(1,:));
ylabel('North error estimate {\it{e_N(t)}}');
nexttile;
plot_estimator(tvec,xhatu(iy,:),Pu(iy,iy,:),x_A(iy,:),'error',z_A(2,:));
ylabel('East error estimate {\it{e_E(t)}}');
title(ti(2),'(a): Baseline Kalman Filter for aircraft A','fontweight','bold');


%% Part b): same as Part a), but design 95% measurement gate
%goal: KF to track/localize aircraft A with measurement gate
%
%use model DTsys_A for this problem
%use measurements z_A for this problem
%
%variables to generate for display/plots:
% Nrej: number of measurement rejections (subset of total nk)
% Trej: times [tvec(k+1)] when msmt is rejected 
% Erej: error [z(:,k+1)-H*x_A(:,k+1)] when msmt is rejected

%msmt gating parameters
Pgate = 0.95;alpha = 1-Pgate;
Lam0  = chi2inv(Pgate, nz);
Nrej=0;Irej=[];

F=DTsys_A.A;G=DTsys_A.B;H=H_A;
nx=length(F);
Q=Q_A;R=R_A;z=z_A;

%%YOUR CODE HERE

%uncomment lines below to output the percent of msmts rejected
% disp('(b) percent of msmts rejected:');
% disp(Nrej/nk*100)

%Generate plot of East and North position error estimates and 2-sigma bounds
%assumes t is (1 x nk), xhatu is (nx x nk), Pu is (nx x nx x nk), xtrue is (nx x nk) 
figs(3)=figure;
ti(3)=tiledlayout(1,2,'TileSpacing','compact','Padding','tight');
nexttile;
plot_estimator(tvec,xhatu(ix,:),Pu(ix,ix,:),x_A(ix,:),'error',z_A(1,:));
ylabel('North error estimate {\it{e_N(t)}}');
hold on
%uncomment line below to add msmt rejections to the plot
%plot(Trej,Erej(1,:),'mo','DisplayName','Msmt Rejection');
hold off;
nexttile;
plot_estimator(tvec,xhatu(iy,:),Pu(iy,iy,:),x_A(iy,:),'error',z_A(2,:));
ylabel('East error estimate {\it{e_E(t)}}');
hold on
%uncomment line below to add msmt rejections to the plot
%plot(Trej,Erej(2,:),'mo','DisplayName','Msmt Rejection');
hold off;
title(ti(3),'(b): Kalman Filter with 95% msmt gating for aircraft A','fontweight','bold');


%% Part c): Estimate Process Noise Covariance Q to make KF consistent
%goal: use hypothesis tests to tune process noise covariance Q 
%
%use model DTsys_A for this problem
%load simulated measurements from tile (these are for aircraft A, but labeled z_c for Part c
%use measurements z_c for this problem
%use states x_c for the plots on problem
%
%variables to generate for display/plots:
% Lam: innovations test statistic lambda (1 x nk)
% LamF: Kalman Filter test statistic lambda (1 x nk)
% Nrej: number of measurement rejections (subset of total nk)
% Trej: times [tvec(k+1)] when msmt is rejected 
% Erej: error [z(:,k+1)-H*x_c(:,k+1)] when msmt is rejected
% NFrej: number of times the filter is inconsistent (subset of total nk)
% TFrej: times [tvec(k+1)] when the filter is inconsistent 
% EFrej: error [z(:,k+1)-H*x_c(:,k+1)] when the filter is inconsistent
%
%load measurement vector z_c and state x_c 
% for aircraft A, simulated with different Q
load ACdata 
%

F=DTsys_A.A;G=DTsys_A.B;H=H_A;
nx=length(F);
Q=Q_A;R=R_A;z=z_c;

%msmt gating parameters
Pgate = 0.95;alpha = 1-Pgate;
Lam0  = chi2inv(Pgate, nz);

%filter consistency parameters
PF = 0.95;alpha = 1-PF;
win=10;%time window
Blow=chi2inv(alpha/2,10*nz)/win;  %low filter threshold
Bhigh=chi2inv(1-alpha/2,10*nz)/win;  %high filter threshold
Lam=zeros(nk,1);
LamF=zeros(nk,1);

%%YOUR CODE HERE

%
%uncomment lines below to output the percent of msmts rejected
% disp('(c) percent of msmts rejected:');
% disp(Nrej/nk*100)
%uncomment lines below to output the percent of time filter is inconsistent
% disp('(c) percent of time filter is inconsistent:');
% disp(NFrej/nk*100)
%
%Generate plot of East and North position error estimates and 2-sigma bounds
%assumes t is (1 x nk), xhatu is (nx x nk), Pu is (nx x nx x nk), xtrue is (nx x nk) 
figs(4)=figure;
ti(4)=tiledlayout(1,2,'TileSpacing','compact','Padding','tight');
nexttile;
plot_estimator(tvec,xhatu(ix,:),Pu(ix,ix,:),x_c(ix,:),'error',z_c(1,:));
ylabel('North error estimate {\it{e_N(t)}}');
hold on
%uncomment the line below to add msmt rejections to the plot
plot(Trej,Erej(1,:),'mo','DisplayName','Msmt Rejection');
hold off;
nexttile;
plot_estimator(tvec,xhatu(iy,:),Pu(iy,iy,:),x_c(iy,:),'error',z_c(2,:));
ylabel('East error estimate {\it{e_E(t)}}');
hold on
%uncomment the line below to add msmt rejections to the plot
plot(Trej,Erej(2,:),'mo','DisplayName','Msmt Rejection');
hold off;
title(ti(4),'(c): Kalman Filter with Msmt Gate and Filter Integrity for aircraft A','fontweight','bold');

figs(5)=figure('Position',[100 100 800 600]);
plot(tvec,Lam,':','Color',MCcolors.mag);
hold on;
plot(tvec((win+1):end),LamF((win+1):end),'-','Color',MCcolors.purple);
plot(tvec,Blow*ones(1,nk),'--','Color',MCcolors.blue);
plot(tvec,Bhigh*ones(1,nk),'--','Color',MCcolors.blue);
hold off;
ylabel('filter integrity')
xlabel('time \it{t} (sec)')
title('(c) Kalman Filter integrity test statistic')
legend('inn test statistic \lambda','KF test statistic \lambda_{10}^{KF}','lower bound','upper bound')


%% Part d): One joint KF for aircraft A and aircraft B
%goal: estimate X=[x_A;x_B] given Z=[z_A;z_B]
%
%continuous time model for aircraft B
Omega_B = -0.045; %turn velocity for V in rad/s
%state vector: [East pos, East vel, North pos, North vel]
A_B=[0 1 0 0;0 0 0 -Omega_B;0 0 0 1;0 Omega_B 0 0]; 
%input forces (East, North)
B_B=[0 0;1 0;0 0;0 1]; 
%discretize 
[F_B,G_B]=c2d(A_B,B_B,dt);
% sensor output models: 2D position states
H_B = [1 0 0 0; 
       0 0 1 0];
%model directional wind disturbances
Q_B = 10*[2.0 0.05;    
         0.05 0.5];
%
% build discrete time model for AC A
DTsys_B = ss(F_B,G_B,H_B,zeros(nz,nw),dt); 

%simulate aircraft B with process noise 
x0_B = [4000,85*cos(pi/4),3200,-85*sin(pi/4)]';
w_B = sqrtm(Q_B)*randn(nw,nk);
[~,~,x_B] = lsim(DTsys_B,w_B,tvec,x0_B);x_B=x_B'; %make into row vector
z_B_clean = H_B * x_B;

%simulate measurements for aircraft A from a tracking station
R_B = [20 0.05; 
      0.05 20];
v_B = sqrtm(R_B)*randn(nz,nk);
z_B = z_B_clean+v_B;
%

%plot trajectories for both aircraft A and B for some context
figs(6)=figure('Position',[100 100 800 600]);
plot_openloop_2Daircraft(x_A,x_B);
title('2D trajectory for aircraft A and B')
%
%use both models DTsys_A,DTsys_B for this problem
%use measurements z_A,z_B for this problem
%
F=blkdiag(DTsys_A.A,DTsys_B.A);
G=blkdiag(DTsys_A.B,DTsys_B.B);
H=blkdiag(H_A,H_B);
nX=length(F);
Q=blkdiag(Q_A,Q_B);
R=blkdiag(R_A,R_B);
Z=[z_A;z_B];

%%YOUR CODE HERE

%Generate plot of East and North position error estimates and 2-sigma bounds
%assumes t is (1 x nk), xhatu is (nx x nk), Pu is (nx x nx x nk), xtrue is (nx x nk) 
figs(7)=figure;
ti(7)=tiledlayout(1,2,'TileSpacing','compact','Padding','tight');
nexttile;
plot_estimator(tvec,xhatu(ix,:),Pu(ix,ix,:),x_A(ix,:),'error',z_A(1,:));
ylabel('North error estimate {\it{e_N(t)}}');
nexttile;
plot_estimator(tvec,xhatu(iy,:),Pu(iy,iy,:),x_A(iy,:),'error',z_A(2,:));
ylabel('East error estimate {\it{e_E(t)}}');
title(ti(7),'(d): Joint Kalman Filter: aircraft A errors','fontweight','bold');
figs(8)=figure;
ti(8)=tiledlayout(1,2,'TileSpacing','compact','Padding','tight');
nexttile;
plot_estimator(tvec,xhatu(nx+ix,:),Pu(nx+ix,nx+ix,:),x_B(ix,:),'error',z_B(1,:));
ylabel('North error estimate {\it{e_N(t)}}');
nexttile;
plot_estimator(tvec,xhatu(nx+iy,:),Pu(nx+iy,nx+iy,:),x_B(iy,:),'error',z_B(2,:));
ylabel('East error estimate {\it{e_E(t)}}');
title(ti(8),'(d): Joint Kalman Filter: aircraft B errors','fontweight','bold');


%% Part e) same as Part d), but uses a 2D relative range sensor
%goal: estimate X=[X_A;X_B] given Z=[z_A;z_B]
%      --> repeat (d), but for a 2D range measurement
%
%use both models DTsys_A,DTsys_B for this problem
%use measurements Zr for this problem
%
Rr = [10 0.15; 
      0.15 10];
Hr = [1 0 0 0 -1 0  0 0;
         0 0 1 0  0 0 -1 0]; 
vr = sqrtm(Rr)*randn(2,nk);
Zr = [x_A(1,:)-x_B(1,:);x_A(3,:)-x_B(3,:)]+vr;

%use both DTsys_A and DTsys_B for this problem
F=blkdiag(DTsys_A.A,DTsys_B.A);
G=blkdiag(DTsys_A.B,DTsys_B.B);
H=Hr;
nX=length(F);
Q=blkdiag(Q_A,Q_B);
R=Rr;
Z=Zr;

%%YOUR CODE HERE

%Generate plot of East and North position error estimates and 2-sigma bounds
%assumes t is (1 x nk), xhatu is (nx x nk), Pu is (nx x nx x nk), xtrue is (nx x nk) 
figs(9)=figure;
ti(9)=tiledlayout(1,2,'TileSpacing','compact','Padding','tight');
nexttile;
plot_estimator(tvec,xhatu(ix,:),Pu(ix,ix,:),x_A(ix,:),'error',z_A(1,:));
ylabel('North error estimate {\it{e_N(t)}}');
nexttile;
plot_estimator(tvec,xhatu(iy,:),Pu(iy,iy,:),x_A(iy,:),'error',z_A(2,:));
ylabel('East error estimate {\it{e_E(t)}}');
title(ti(10),'(e): Joint KF with Relative Range: aircraft A errors','fontweight','bold');
figs(10)=figure;
ti(10)=tiledlayout(1,2,'TileSpacing','compact','Padding','tight');
nexttile;
plot_estimator(tvec,xhatu(nx+ix,:),Pu(nx+ix,nx+ix,:),x_B(ix,:),'error',z_B(1,:));
ylabel('North error estimate {\it{e_N(t)}}');
nexttile;
plot_estimator(tvec,xhatu(nx+iy,:),Pu(nx+iy,nx+iy,:),x_B(iy,:),'error',z_B(2,:));
ylabel('East error estimate {\it{e_E(t)}}');
title(ti(10),'(e): Joint KF with Relative Range: aircraft B errors','fontweight','bold');


%% Plotting functions
function plot_openloop_2Daircraft(x_A,x_B);
% x_A is the state of aircraft A [East pos, East vel, North pos, North vel]
% optional: x_B is the state of aircraft B [East pos, East vel, North pos, North vel]
%
global MCcolors;
%plot aircraft A
p1=plot(x_A(1,:),x_A(3,:),'-','color',MCcolors.blue);
hold on;
plot(x_A(1,1),x_A(3,1),'b>','markersize',10);
plot(x_A(1,end),x_A(3,end),'pentagram','Color','b','markersize',10);
hold off
ylabel('North');xlabel('East');
grid;
%
if nargin<2, %only aircraft A  
    legend([p1],'aircraft A');
else, %overlay aircraft B
    hold on;
    p2=plot(x_B(1,:),x_B(3,:),':','color',MCcolors.red);
    hold on;
    plot(x_B(1,1),x_B(3,1),'>','color',MCcolors.red,'markersize',10);
    plot(x_B(1,end),x_B(3,end),'pentagram','Color',MCcolors.red,'markersize',10);
    hold off
    legend([p1 p2],'aircraft A','aircraft B');
end
end

%% --------- start-up items
function MAE6760startup(font_size);
%
% define colors for plotting
global MCcolors; %define colors as global with access 
MCcolors.red=[200,0,0]/255;
MCcolors.blue=[4,51,255]/255;
MCcolors.purple=[147,23,255]/255;
MCcolors.green=[0,160,0]/255;
MCcolors.orange=[253,128,8]/255;
MCcolors.mag=[255,64,255]/255;
MCcolors.cyan=[0,230,255]/255;
%
% define standard figure positioning and size
set(groot,'DefaultFigureUnits','pixels');
set(groot,'DefaultFigurePosition',[100 100 1600 600]);
set(groot,'DefaultFigureWindowStyle','normal');  % Important
set(groot,'DefaultAxesFontSize',16);
set(groot,'DefaultAxesFontWeight','bold');
set(groot,'DefaultLineLineWidth',2);
%
end